main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala

```diff
@@ -0,0 +1,208 @@
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.language.ast.shared.SymUse.{CaseSymUse, EffectSymUse, OpSymUse}
+import ca.uwaterloo.flix.language.ast.shared.*
+import ca.uwaterloo.flix.util.collection.Nel
+
+object OccurrenceAst {
+
+  case class Root(defs: Map[Symbol.DefnSym, Def],
+                  enums: Map[Symbol.EnumSym, MonoAst.Enum],
+                  structs: Map[Symbol.StructSym, MonoAst.Struct],
+                  effects: Map[Symbol.EffectSym, MonoAst.Effect],
+                  mainEntryPoint: Option[Symbol.DefnSym],
+                  entryPoints: Set[Symbol.DefnSym],
+                  sources: Map[Source, SourceLocation])
+
+  case class Def(sym: Symbol.DefnSym, fparams: List[(MonoAst.FormalParam, Occur)], spec: MonoAst.Spec, exp: Expr, context: DefContext, loc: SourceLocation)
+
+  sealed trait Expr extends Product {
+    def tpe: Type
+
+    def eff: Type
+
+    def loc: SourceLocation
+  }
+
+  object Expr {
+
+    case class Cst(cst: Constant, tpe: Type, loc: SourceLocation) extends Expr {
+      def eff: Type = Type.Pure
+    }
+
+    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends Expr {
+      def eff: Type = Type.Pure
+    }
+
+    case class Lambda(fparam: (MonoAst.FormalParam, Occur), exp: Expr, tpe: Type, loc: SourceLocation) extends Expr {
+      def eff: Type = Type.Pure
+    }
+
+    case class ApplyAtomic(op: AtomicOp, exps: List[Expr], tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class ApplyClo(exp1: Expr, exp2: Expr, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class ApplyDef(sym: Symbol.DefnSym, exps: List[Expr], itpe: Type, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class ApplyLocalDef(sym: Symbol.VarSym, exps: List[Expr], tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class Let(sym: Symbol.VarSym, exp1: Expr, exp2: Expr, tpe: Type, eff: Type, occur: Occur, loc: SourceLocation) extends Expr
+
+    case class LocalDef(sym: Symbol.VarSym, fparams: List[(MonoAst.FormalParam, Occur)], exp1: Expr, exp2: Expr, tpe: Type, eff: Type, occur: Occur, loc: SourceLocation) extends Expr
+
+    case class Scope(sym: Symbol.VarSym, regSym: Symbol.RegionSym, exp: Expr, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class IfThenElse(exp1: Expr, exp2: Expr, exp3: Expr, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class Stm(exp1: Expr, exp2: Expr, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class Discard(exp: Expr, eff: Type, loc: SourceLocation) extends Expr {
+      def tpe: Type = Type.mkUnit(loc)
+    }
+
+    case class Match(exp: Expr, rules: List[MatchRule], tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+
+    case class VectorLit(exps: List[Expr], tpe: Type, eff: Type, loc: SourceLocation) extends Expr
```

Comment:
What's the reason Vectors are here but not atomic?

Reply:
I don't understand. `ApplyAtomic` is early in the file. `OccurrenceAst` has the same expressions as `MonoAst`.